pkg/coscheduling/core/core.go

```diff
@@ -0,0 +1,292 @@
+		// So here it just serves as a sanity check.
+		return false, fmt.Errorf("PodGroup not found")
+	}
+	if pgStatus.PodGroup.Status.Phase == "" || pgStatus.PodGroup.Status.Phase == v1alpha1.PodGroupPending {
+		pgStatus.PodGroup.Status.Phase = v1alpha1.PodGroupPreScheduling
+	}
+
+	bound := pgMgr.calculateBoundPods(pgStatus.PodGroup.Name, pgStatus.PodGroup.Namespace)
+	// The bound is calculated from the snapshot. The current pod does not exist in the snapshot during this scheduling cycle.
+	ready := int32(bound)+1 >= pgStatus.PodGroup.Spec.MinMember
+	if ready {
+		pgStatus.Scheduled = true
+		return true, nil
+	}
+	return false, util.ErrorWaiting
+}
+
+// PostBind is used for send metrics and update to api if all finish
+func (pgMgr *PodGroupManager) PostBind(ctx context.Context, pod *corev1.Pod, nodeName string) {
+	pgMgr.Lock()
+	defer pgMgr.Unlock()
+	pgFullName, pgStatus := pgMgr.GetPodGroup(pod)
+	if pgFullName == "" || pgStatus == nil {
+		return
+	}
+	pgCopy := pgStatus.PodGroup.DeepCopy()
+	pgCopy.Status.Scheduled++
+
+	if pgCopy.Status.Scheduled >= pgStatus.PodGroup.Spec.MinMember {
+		pgCopy.Status.Phase = v1alpha1.PodGroupScheduled
+	} else {
+		pgCopy.Status.Phase = v1alpha1.PodGroupScheduling
+		if pgCopy.Status.ScheduleStartTime.IsZero() {
+			pgCopy.Status.ScheduleStartTime = metav1.Time{Time: time.Now()}
+		}
+	}
+	pgStatus.PodGroup.Status = pgCopy.Status
+	if pgCopy.Status.Phase != pgStatus.PodGroup.Status.Phase {
+
+		pg, err := pgMgr.pgLister.PodGroups(pgCopy.Namespace).Get(pgCopy.Name)
+		if err != nil {
+			klog.Error(err)
+			return
+		}
+		patch, err := util.CreateMergePatch(pg, pgCopy)
+		if err != nil {
+			klog.Error(err)
+			return
+		}
+		if err := pgMgr.PatchPodGroup(pg.Name, pg.Namespace, patch); err != nil {
+			klog.Error(err)
+			return
+		}
+		pgStatus.PodGroup.Status.Phase = pgCopy.Status.Phase
+	}
+	pgStatus.PodGroup.Status.Scheduled = pgCopy.Status.Scheduled
+	return
+}
+
+// getCreationTime return creation time of podgroup or pod
+func (pgMgr *PodGroupManager) GetCreationTime(pod *corev1.Pod, ts time.Time) time.Time {
+	pgName := util.GetPodGroupLabel(pod)
+	if len(pgName) == 0 {
+		return ts
+	}
+	pg, err := pgMgr.pgLister.PodGroups(pod.Namespace).Get(pgName)
+	if err != nil {
+		return ts
+	}
+	return pg.CreationTimestamp.Time
+}
+
+// AddToDenyCache add podGroup to the back list cache
+func (pgMgr *PodGroupManager) AddToDenyCache(pgFullName string) {
+	pgMgr.lastDeniedPG.Add(pgFullName, "", 20*time.Second)
+}
+
+// PatchPodGroup
+func (pgMgr *PodGroupManager) PatchPodGroup(pgName string, namespace string, patch []byte) error {
+	_, err := pgMgr.pgClient.SchedulingV1alpha1().PodGroups(namespace).Patch(context.TODO(), pgName,
+		types.MergePatchType, patch, metav1.PatchOptions{})
+	return err
+}
+
+// GetPodGroup get PodGroup from cache
+func (pgMgr *PodGroupManager) GetPodGroup(pod *corev1.Pod) (string, *cache.PodGroupSchedulingStatus) {
```

Comment:
Maintaining a PGSchedulingStatus is not that necessary now. We can directly return the PodGroup object from pgLister.